### DISCUSSION TITLE
Allure Report in C# selenium framework

### DISCUSSION THREAD

**Discussion by RachitaDangare** [allure-framework/allure2#2540] [Questions & Support]
How we can call a method from [OneTimeSetUp] which has allure step associated with it. 
When I do this I am getting bellow exception. 
OneTimeSetUp: System.InvalidOperationException : No step context is active.

Code Snippet for reference -

[SetUpFixture]
public class OneTimeFixture
{
        [OneTimeSetUp]
        public static void StartUpAssembly()
        {
           ABC abc = new ABC();
             abc.Test()
        }
} 

public class ABC
{
     [Step("Log details")]
      public void Test()
      {
       // Does something
      }
}

Could you please suggest how we can call Allure step associated method call from OneTimeSetUp method.

**Comment by delatrie** [comment]
Currently, `[OneTimeSetUp]` is only supported for test fixtures (i.e., test classes). It also needs `[AllureBefore]` to be applied to the method:

```csharp
using NUnit.Framework;
using Allure.NUnit;
using Allure.NUnit.Attributes;

[AllureNUnit]
public class MyTestClass
{
    [OneTimeSetUp]
    [AllureBefore]
    public void StartUpAssembly()
    {
        ABC abc = new ABC();
        abc.Test();
    }
}
```

The feature request for `[SetupFixture]` is here: allure-framework/allure-csharp#361.

**Reply by RachitaDangare** [reply]
Hi @delatrie , Thank you for quick response. 

After using [AllureBefore] I am getting below error

 OneTimeSetUp: System.InvalidOperationException : No fixture context is active.

**Reply by delatrie** [reply]
`[AllureNUnit]` is missing in my example. It should be like this (I've fixed the example in my comment above):

```csharp
using NUnit.Framework;
using Allure.NUnit;
using Allure.NUnit.Attributes;

[AllureNUnit] // <- note the attribute
public class MyTestClass
{
    [OneTimeSetUp]
    [AllureBefore]
    public void StartUpAssembly()
    {
        ABC abc = new ABC();
        abc.Test();
    }
}
```

Be careful as this is a workaround that differs from `[SetupFixture]` behavior. If you have multiple test classes inherited from that base, and you need to ensure `abc.Test()` is only called once, a guard variable will help you:

```csharp
using NUnit.Framework;
using Allure.NUnit;
using Allure.NUnit.Attributes;

[AllureNUnit]
public class MyTestClass
{
    private static bool isInitialized = false;

    [OneTimeSetUp]
    [AllureBefore]
    public void StartUpAssembly()
    {
        if (!isInitialized)
        {
            ABC abc = new ABC();
            abc.Test();
            isInitialized = true;
        }
    }
}
```

**Reply by RachitaDangare** [reply]
Hi @delatrie, I have tried to use guard variable to make sure Test() method called once for given test suite. But its not working as expected.
I mean Test() method is getting called for each test. 

This Test() method is getting called even we execute multiple tests from single test class file or we have tests from multiple test classes. 

I think by using [AllureBefore] attribute with [OneTimeSetUp]. Its deviating actual OneTimeSetUp attribute behavior.  And its not getting fixed by using guard variable as well. 

Could you please help me with this? I don't want to change the OneTimeSetUp method behavior , I expect it should call once before test execution starts.
